Additional logging when endpoint changes.

winksaville opened the pull request:
This adds logging so the user is informed whenever the endpoint changes.

beacon_node/eth1/Cargo.toml:
 * Added test_logger features

beacon_node/eth1/src/inner.rs:
 * Added endpoint_index to Inner to track the current endpoint index
 * Added fn update_endpoint_index which updates endpoint_index and
   returns the old value.

beacon_node/eth1/src/service.rs:
 * Changed num_errors to cur_index
 * Changed to logging to output different messages based on when
   the endpoint changed.

beacon_node/eth1/tests/test.rs
 * Added fn test_logger extracted from
   beacon_node/beacon_chain/src/test_utils.rs
 * Use test_logger instead of null_logger in test_fallback_when_offline

Output of test_fallback_when_offline with test_logger feature enabled:

  wink@3900x:~/prgs/ethereum/myrepos/lighthouse/beacon_node/eth1 (Additional-logging-when-endpoint-changes-take2)
  $ cargo test test_fallback_when_offline --features test_logger -- --nocapture
      Finished test [unoptimized + debuginfo] target(s) in 0.19s
       Running unittests (/home/wink/prgs/ethereum/myrepos/lighthouse/target/debug/deps/eth1-209b6ecb1bd266e0)

  running 0 tests

  test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 11 filtered out; finished in 0.00s

       Running tests/test.rs (/home/wink/prgs/ethereum/myrepos/lighthouse/target/debug/deps/test-5ed8c8c2ba222544)

  running 1 test
  Sep 09 14:49:28.039 INFO First endpoint is preferred, url: http://localhost:45637/, index: 0, module: eth1::service:755
  Sep 09 14:49:28.046 DEBG Downloading eth1 blocks, last: None, first: None, module: eth1::service:1134
  Sep 09 14:49:28.046 DEBG No new eth1 blocks imported, cached_blocks: 0, latest_block: None, module: eth1::service:1206
  Sep 09 14:49:28.100 DEBG Imported deposit logs chunk, logs: 0, module: eth1::service:1016
  Sep 09 14:49:28.100 DEBG No new deposits found, total_deposits: 0, latest_block: None, module: eth1::service:1040
  Sep 09 14:49:28.139 INFO Endoint changed to fallback, url: http://localhost:46809/, index: 1, module: eth1::service:745
  Sep 09 14:49:28.145 DEBG Downloading eth1 blocks, last: Some(6), first: Some(1), module: eth1::service:1134
  Sep 09 14:49:28.146 WARN Error connecting to eth1 node endpoint, action: trying fallbacks, endpoint: http://localhost:45637/, module: eth1::service:187
  Sep 09 14:49:28.158 DEBG Imported deposit logs chunk, logs: 0, module: eth1::service:1016
  Sep 09 14:49:28.158 DEBG No new deposits found, total_deposits: 0, latest_block: None, module: eth1::service:1040
  Sep 09 14:49:28.197 DEBG Imported eth1 block(s), new: 6, total_cached_blocks: 6, latest_block: 6, latest_block_age: 0 mins, module: eth1::service:1197
  test fallbacks::test_fallback_when_offline ... ok

  test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 13 filtered out; finished in 2.95s


Output when test_logger feature is NOT enabled:

  wink@3900x:~/prgs/ethereum/myrepos/lighthouse/beacon_node/eth1 (Additional-logging-when-endpoint-changes-take2)
  $ cargo test test_fallback_when_offline -- --nocapture
     Compiling eth1 v0.2.0 (/home/wink/prgs/ethereum/myrepos/lighthouse/beacon_node/eth1)
      Finished test [unoptimized + debuginfo] target(s) in 9.92s
       Running unittests (/home/wink/prgs/ethereum/myrepos/lighthouse/target/debug/deps/eth1-675a42b60865a619)

  running 0 tests

  test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 11 filtered out; finished in 0.00s

       Running tests/test.rs (/home/wink/prgs/ethereum/myrepos/lighthouse/target/debug/deps/test-d4f896e20deed138)

  running 1 test
  test fallbacks::test_fallback_when_offline ... ok

  test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 13 filtered out; finished in 3.06s

winksaville committed:
Additional logging when endpoint changes.

This adds logging so the user is informed whenever the endpoint changes.
If the old endpoint is None you'll see the idx and url only of the new one:new idx and url and old will be None:

  Sep 20 14:23:42.312 INFO New endpoint: idx=0 url=http://localhost:45757/ Old endpoint: None, module: eth1::service:745

If the endpoint is changing you'll see the idx and url of both new and old:

    Sep 20 14:23:42.423 INFO New endpoint: idx=1 url=http://localhost:43889/ Old endpoint: idx=0 url=http://localhost:45757/, module: eth1::service:738


beacon_node/eth1/Cargo.toml:
 * Depend on logging packange

beacon_node/eth1/src/inner.rs:
 * Added endpoint_index to Inner to track the current endpoint index
 * Added fn update_endpoint_index which updates endpoint_index and
   returns the previous value.

beacon_node/eth1/src/service.rs:
 * Changed num_errors to cur_index
 * Changed to logging to output different messages based on when
   the endpoint changed.

beacon_node/eth1/tests/test.rs
 * Use test_logger instead of null_logger in test_fallback_when_offline



Here is the output of test_fallback_when_offline with test_logger
feature enabled:

  wink@3900x:~/prgs/ethereum/myrepos/lighthouse (Additional-logging-when-endpoint-changes)
  $ cargo test -p eth1 fallbacks::test_fallback_when_offline  --features "logging/test_logger"
      Finished test [unoptimized + debuginfo] target(s) in 0.19s
       Running unittests (target/debug/deps/eth1-e41f5be9cafe96b8)

  running 0 tests

  test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 11 filtered out; finished in 0.00s

       Running tests/test.rs (target/debug/deps/test-d9ded8d7d852e1be)

  running 1 test
  Sep 20 14:23:42.312 INFO New endpoint: idx=0 url=http://localhost:45757/ Old endpoint: None, module: eth1::service:745
  Sep 20 14:23:42.319 DEBG Downloading eth1 blocks, last: None, first: None, module: eth1::service:1114
  Sep 20 14:23:42.319 DEBG No new eth1 blocks imported, cached_blocks: 0, latest_block: None, module: eth1::service:1186
  Sep 20 14:23:42.372 DEBG Imported deposit logs chunk, logs: 0, module: eth1::service:996
  Sep 20 14:23:42.372 DEBG No new deposits found, total_deposits: 0, latest_block: None, module: eth1::service:1020
  Sep 20 14:23:42.423 INFO New endpoint: idx=1 url=http://localhost:43889/ Old endpoint: idx=0 url=http://localhost:45757/, module: eth1::service:738
  Sep 20 14:23:42.430 DEBG Downloading eth1 blocks, last: Some(6), first: Some(1), module: eth1::service:1114
  Sep 20 14:23:42.431 WARN Error connecting to eth1 node endpoint, action: trying fallbacks, endpoint: http://localhost:45757/, module: eth1::service:187
  Sep 20 14:23:42.448 DEBG Imported deposit logs chunk, logs: 0, module: eth1::service:996
  Sep 20 14:23:42.448 DEBG No new deposits found, total_deposits: 0, latest_block: None, module: eth1::service:1020
  Sep 20 14:23:42.498 DEBG Imported eth1 block(s), new: 6, total_cached_blocks: 6, latest_block: 6, latest_block_age: 0 mins, module: eth1::service:1177
  test fallbacks::test_fallback_when_offline ... ok

  test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 13 filtered out; finished in 2.99s

  wink@3900x:~/prgs/ethereum/myrepos/lighthouse (Additional-logging-when-endpoint-changes)



And here it is when test_logger feature is NOT enabled:

    wink@3900x:~/prgs/ethereum/myrepos/lighthouse (Additional-logging-when-endpoint-changes)
  $ cargo test -p eth1 fallbacks::test_fallback_when_offline
     Compiling eth1 v0.2.0 (/home/wink/prgs/ethereum/myrepos/lighthouse/beacon_node/eth1)
      Finished test [unoptimized + debuginfo] target(s) in 10.87s
       Running unittests (target/debug/deps/eth1-8a8a5b39332fc92a)

  running 0 tests

  test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 11 filtered out; finished in 0.00s

       Running tests/test.rs (target/debug/deps/test-730e97e5b27aeb01)

  running 1 test
  test fallbacks::test_fallback_when_offline ... ok

  test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 13 filtered out; finished in 2.96s

winksaville commented:
Forced push fix for failing `cargo fmt`

winksaville commented:
Rebased on top of #2586 and simplified the logging output.

winksaville commented:
Rebased on top of #2586 and is ready for review/merge after #2586 is merged.

winksaville commented:
Rebased on latest unstable which now has #2586 merged, this is ready for review.

paulhauner commented on beacon_node/eth1/src/service.rs:
```diff
@@ -730,9 +731,29 @@ impl Service {
+        match prev_index {
+            Some(prev_index) => {
+                if cur_index != prev_index {
+                    info!(
```
It would be more consistent (and perhaps a little easier) to specify these logs in this style:

https://github.com/sigp/lighthouse/blob/7c88f582d955537f7ffff9b2c879dcf5bf80ce13/beacon_node/client/src/notifier.rs#L206-L212

It would also mean we get the correct spacing after `"New endpoint:"`.

paulhauner commented on beacon_node/eth1/src/service.rs:
```diff
@@ -730,9 +731,29 @@ impl Service {
+            None => {
+                info!(
+                    self.log,
+                    "New endpoint: idx={} url={} Old endpoint: None",
```
As above

paulhauner commented on beacon_node/eth1/src/service.rs:
```diff
@@ -730,9 +731,29 @@ impl Service {
+                        self.log,
+                        "New endpoint: idx={} url={} Old endpoint: idx={} url={}",
+                        cur_index,
+                        &endpoints.fallback.servers[cur_index].endpoint.as_ref(),
```
I think it would be best to use a non-panic-y `servers.get()` here. We generally try to avoid direct index access in Lighthouse.

paulhauner commented on beacon_node/eth1/src/service.rs:
```diff
@@ -730,9 +731,29 @@ impl Service {
+                        cur_index,
+                        &endpoints.fallback.servers[cur_index].endpoint.as_ref(),
+                        prev_index,
+                        &endpoints.fallback.servers[prev_index].endpoint.as_ref()
```
As above.

paulhauner commented on beacon_node/eth1/src/service.rs:
```diff
@@ -730,9 +731,29 @@ impl Service {
+                    self.log,
+                    "New endpoint: idx={} url={} Old endpoint: None",
+                    cur_index,
+                    &endpoints.fallback.servers[cur_index].endpoint.as_ref()
```
As above (safe getter).

paulhauner commented on beacon_node/eth1/src/service.rs:
```diff
@@ -730,9 +731,29 @@ impl Service {
+                if cur_index != prev_index {
+                    info!(
+                        self.log,
+                        "New endpoint: idx={} url={} Old endpoint: idx={} url={}",
```
Perhaps we could give more context about the endpoint we're referring to? Perhaps something like "Changing eth1 endpoint" or something like that?

paulhauner reviewed (changes requested):
Thanks for taking this one on! I think the info will be useful to users :relaxed: 

I've left some comments in the current code, but I think some further changes will be needed. This code works well for the `get_remote_head_and_new_block_ranges` function, however we make further calls to the eth1 node that may involve switching endpoints.

For example, after running `get_remote_head_and_new_block_ranges` we may run this call which can also involve switching between fallbacks:

https://github.com/sigp/lighthouse/blob/7c88f582d955537f7ffff9b2c879dcf5bf80ce13/beacon_node/eth1/src/service.rs#L890-L896

So, I think we need to boost your logic up higher the call stack. To that end, I think the best place for this logic would be inside `EndpointsCache::first_success`. Specifically, here:

https://github.com/sigp/lighthouse/blob/7c88f582d955537f7ffff9b2c879dcf5bf80ce13/beacon_node/eth1/src/service.rs#L145

I suggest we store the latest `endpoint.endpoint` (much like your `endpoint_index`) and then create a log whenever we hit that `Ok(t)` branch with a new `endpoint.endpoint` value. That should cover a broader range of cases.

winksaville committed:
Cleanup suggestions by Paul Hauner

Step 1 have both my cur_index in inner and cur_endpoint_url.

winksaville committed:
Remove endpoint_index

We are now using just EndpointsCache::cur_endpoint_url.

The out from testing is:
```
wink@3900x:~/prgs/ethereum/myrepos/lighthouse (Additional-logging-when-endpoint-changes)
$ cargo test -p eth1 test_fallback_when_offline --features "logging/test_logger" -- --nocapture
    Finished test [unoptimized + debuginfo] target(s) in 0.18s
     Running unittests (target/debug/deps/eth1-6ef59c9498f99781)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 11 filtered out; finished in 0.00s

     Running tests/test.rs (target/debug/deps/test-c359e1dbd62c3a15)

running 1 test
Oct 11 17:01:22.588 INFO Changing to eth1 endpoint http://localhost:35783/, module: eth1::service:164
Oct 11 17:01:22.588 INFO             from endpoint None, module: eth1::service:169
Oct 11 17:01:22.596 DEBG Downloading eth1 blocks, last: None, first: None, module: eth1::service:1132
Oct 11 17:01:22.596 DEBG No new eth1 blocks imported, cached_blocks: 0, latest_block: None, module: eth1::service:1204
Oct 11 17:01:22.645 DEBG Imported deposit logs chunk, logs: 0, module: eth1::service:1014
Oct 11 17:01:22.645 DEBG No new deposits found, total_deposits: 0, latest_block: None, module: eth1::service:1038
Oct 11 17:01:22.705 INFO Changing to eth1 endpoint http://localhost:36993/, module: eth1::service:164
Oct 11 17:01:22.705 INFO             from endpoint http://localhost:35783/, module: eth1::service:169
Oct 11 17:01:22.712 DEBG Downloading eth1 blocks, last: Some(6), first: Some(1), module: eth1::service:1132
Oct 11 17:01:22.713 WARN Error connecting to eth1 node endpoint, action: trying fallbacks, endpoint: http://localhost:35783/, module: eth1::service:222
Oct 11 17:01:22.731 DEBG Imported deposit logs chunk, logs: 0, module: eth1::service:1014
Oct 11 17:01:22.731 DEBG No new deposits found, total_deposits: 0, latest_block: None, module: eth1::service:1038
Oct 11 17:01:22.784 DEBG Imported eth1 block(s), new: 6, total_cached_blocks: 6, latest_block: 6, latest_block_age: 0 mins, module: eth1::service:1195
test fallbacks::test_fallback_when_offline ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 13 filtered out; finished in 2.94s
```

winksaville commented:
> I suggest we store the latest `endpoint.endpoint` (much like your `endpoint_index`) and then create a log whenever we hit that `Ok(t)` branch with a new `endpoint.endpoint` value. That should cover a broader range of cases.

@paulhauner I believe I done what you've requested, although there are likely misunderstandings on my part :) Please take another look when you have time. The output is now:
```
wink@3900x:~/prgs/ethereum/myrepos/lighthouse (Additional-logging-when-endpoint-changes)
$ cargo test -p eth1 test_fallback_when_offline --features "logging/test_logger" -- --nocapture
    Finished test [unoptimized + debuginfo] target(s) in 0.18s
     Running unittests (target/debug/deps/eth1-6ef59c9498f99781)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 11 filtered out; finished in 0.00s

     Running tests/test.rs (target/debug/deps/test-c359e1dbd62c3a15)

running 1 test
Oct 11 17:01:22.588 INFO Changing to eth1 endpoint http://localhost:35783/, module: eth1::service:164
Oct 11 17:01:22.588 INFO             from endpoint None, module: eth1::service:169
Oct 11 17:01:22.596 DEBG Downloading eth1 blocks, last: None, first: None, module: eth1::service:1132
Oct 11 17:01:22.596 DEBG No new eth1 blocks imported, cached_blocks: 0, latest_block: None, module: eth1::service:1204
Oct 11 17:01:22.645 DEBG Imported deposit logs chunk, logs: 0, module: eth1::service:1014
Oct 11 17:01:22.645 DEBG No new deposits found, total_deposits: 0, latest_block: None, module: eth1::service:1038
Oct 11 17:01:22.705 INFO Changing to eth1 endpoint http://localhost:36993/, module: eth1::service:164
Oct 11 17:01:22.705 INFO             from endpoint http://localhost:35783/, module: eth1::service:169
Oct 11 17:01:22.712 DEBG Downloading eth1 blocks, last: Some(6), first: Some(1), module: eth1::service:1132
Oct 11 17:01:22.713 WARN Error connecting to eth1 node endpoint, action: trying fallbacks, endpoint: http://localhost:35783/, module: eth1::service:222
Oct 11 17:01:22.731 DEBG Imported deposit logs chunk, logs: 0, module: eth1::service:1014
Oct 11 17:01:22.731 DEBG No new deposits found, total_deposits: 0, latest_block: None, module: eth1::service:1038
Oct 11 17:01:22.784 DEBG Imported eth1 block(s), new: 6, total_cached_blocks: 6, latest_block: 6, latest_block_age: 0 mins, module: eth1::service:1195
test fallbacks::test_fallback_when_offline ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 13 filtered out; finished in 2.94s
```

paulhauner commented on beacon_node/eth1/src/service.rs:
```diff
@@ -81,6 +81,7 @@ async fn get_state(endpoint: &EndpointWithState) -> Option<EndpointState> {
 /// is not usable.
 pub struct EndpointsCache {
     pub fallback: Fallback<EndpointWithState>,
+    pub cur_endpoint_url: RwLock<Option<SensitiveUrl>>,
```
Since we never call `read` on this `RwLock`, we can use the simpler `Mutex`.

paulhauner commented on beacon_node/eth1/src/service.rs:
```diff
@@ -871,7 +903,7 @@ impl Service {
     ///
     /// Emits logs for debugging and errors.
-    pub async fn update_deposit_cache(
+    pub async fn update_deposit_cache<'e>(
```
I think we should drop this `<'e>`.

paulhauner commented on beacon_node/eth1/src/service.rs:
```diff
@@ -1028,7 +1060,7 @@ impl Service {
     ///
     /// Emits logs for debugging and errors.
-    pub async fn update_block_cache(
+    pub async fn update_block_cache<'e>(
```
As above.

winksaville replied:
I do believe it's desirable to see the startup/initial value so I know what the initial value is, why don't you want it?

Another deseriable capability is to have an api which returns the current value, if that was or is already provided then I could be convinced the initial value isn't needed. What do you think?

Finally, if you rather get some in quickly, we can abandoned this PR and just take yours, https://github.com/paulhauner/lighthouse/commit/d917c5d7650280e02f47380fd084528d1a45a0ac. I'm fine with that!!

paulhauner commented on beacon_node/eth1/src/service.rs:
```diff
@@ -142,7 +157,27 @@ impl EndpointsCache {
+                                let (prev_url, changed) =
+                                    self.update_cur_endpoint_url(endpoint.endpoint.clone());
+                                if changed {
+                                    info!(
```
This is not quite the log format I was referring to. I've added an example of what I was trying to communicate in https://github.com/paulhauner/lighthouse/commit/d917c5d7650280e02f47380fd084528d1a45a0ac.

paulhauner commented on beacon_node/eth1/src/service.rs:
```diff
@@ -121,6 +122,20 @@ impl EndpointsCache {
+        let changed: bool = if let Some(purl) = &prev_url {
+            purl.full != url.full
+        } else {
+            true
```
Here we're going to create an "endpoint changed" log each time we startup. Was that what you had intended? I thought that might be a little noisy but I'm open to it.

Over in https://github.com/paulhauner/lighthouse/commit/d917c5d7650280e02f47380fd084528d1a45a0ac I've made a suggestion for one that won't log the message on startup. Even if you want to log on startup, you might find the `map_or` function useful for the `if let Some(_) = _ { ... }  else { true }` pattern :)

paulhauner reviewed (changes requested):
Yep, this is where I think it's best to call it.

I just have a few minor suggestions. I ended up implementing them all in https://github.com/paulhauner/lighthouse/commit/d917c5d7650280e02f47380fd084528d1a45a0ac because I needed the compiler to help me be correct :sweat_smile:

winksaville commented:
@paulhauner, I'm kinda busy and the long feed back loop is making it hard to contribute. At this point I it appears you've got what you feel is needed. So it seems to me the best approach is to just merge your change, [pauhauner@d917c5d](https://github.com/paulhauner/lighthouse/commit/d917c5d7650280e02f47380fd084528d1a45a0ac). I'll close this PR.
